File: archive/cdk-ts/Pr4347/lib/dynamodb-stack.ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';

export interface DynamoDBStackProps extends cdk.StackProps {
  environmentSuffix: string;
}

export class DynamoDBStack extends cdk.Stack {
  public readonly globalTable: dynamodb.Table;

  constructor(scope: Construct, id: string, props: DynamoDBStackProps) {
    super(scope, id, props);

    const suffix = props.environmentSuffix;

    // Create a DynamoDB Global Table
    this.globalTable = new dynamodb.Table(this, 'TradingGlobalTable', {
      tableName: `trading-transactions-${suffix}`,
      partitionKey: { name: 'id', type: dynamodb.AttributeType.STRING },
      sortKey: { name: 'timestamp', type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      pointInTimeRecovery: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      replicationRegions: ['us-west-2'],
    });

    // Outputs for integration testing
    new cdk.CfnOutput(this, 'GlobalTableArn', {
      value: this.globalTable.tableArn,
      description: 'ARN of the Trading Global Table',
      exportName: `trading-global-table-arn-${suffix}`,
    });

    new cdk.CfnOutput(this, 'GlobalTableName', {
      value: this.globalTable.tableName,
      description: 'Name of the Trading Global Table',
      exportName: `trading-global-table-name-${suffix}`,
    });

    /* istanbul ignore next */
    const streamArn = this.globalTable.tableStreamArn || 'N/A';
    new cdk.CfnOutput(this, 'GlobalTableStreamArn', {
      value: streamArn,
      description: 'Stream ARN of the Trading Global Table',
      exportName: `trading-global-table-stream-arn-${suffix}`,
    });

    new cdk.CfnOutput(this, 'PartitionKeyName', {
      value: 'id',
      description: 'Partition key name of the table',
      exportName: `trading-table-partition-key-${suffix}`,
    });

    new cdk.CfnOutput(this, 'SortKeyName', {
      value: 'timestamp',
      description: 'Sort key name of the table',
      exportName: `trading-table-sort-key-${suffix}`,
    });

    new cdk.CfnOutput(this, 'BillingMode', {
      value: 'PAY_PER_REQUEST',
      description: 'Billing mode of the table',
      exportName: `trading-table-billing-mode-${suffix}`,
    });

    new cdk.CfnOutput(this, 'PointInTimeRecovery', {
      value: 'true',
      description: 'Point-in-time recovery enabled status',
      exportName: `trading-table-pitr-${suffix}`,
    });

    new cdk.CfnOutput(this, 'ReplicationRegions', {
      value: 'us-west-2',
      description: 'Replication regions for the global table',
      exportName: `trading-table-replication-regions-${suffix}`,
    });

    new cdk.CfnOutput(this, 'PrimaryRegion', {
      value: this.region,
      description: 'Primary region for the global table',
      exportName: `trading-table-primary-region-${suffix}`,
    });
  }
}

File: archive/cdk-py/Pr6928/lib/dynamodb_stack.py
"""dynamodb_stack.py
DynamoDB global tables with point-in-time recovery.
"""

from typing import Optional
import aws_cdk as cdk
from aws_cdk import aws_dynamodb as dynamodb
from constructs import Construct


class DynamoDBStackProps:
    """Properties for DynamoDB stack."""

    def __init__(
        self,
        environment_suffix: str,
        primary_region: str,
        secondary_region: str
    ):
        self.environment_suffix = environment_suffix
        self.primary_region = primary_region
        self.secondary_region = secondary_region


class DynamoDBStack(Construct):
    """Creates DynamoDB global table with PITR."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        props: DynamoDBStackProps
    ):
        super().__init__(scope, construct_id)

        # Create DynamoDB table (single region for testing)
        self.table = dynamodb.TableV2(
            self,
            f'TransactionTable{props.environment_suffix}',
            table_name=f'dr-transactions-{props.environment_suffix}',
            partition_key=dynamodb.Attribute(
                name='transactionId',
                type=dynamodb.AttributeType.STRING
            ),
            sort_key=dynamodb.Attribute(
                name='timestamp',
                type=dynamodb.AttributeType.NUMBER
            ),
            billing=dynamodb.Billing.on_demand(),
            point_in_time_recovery=True,
            deletion_protection=False,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            global_secondary_indexes=[
                dynamodb.GlobalSecondaryIndexPropsV2(
                    index_name='StatusIndex',
                    partition_key=dynamodb.Attribute(
                        name='status',
                        type=dynamodb.AttributeType.STRING
                    ),
                    sort_key=dynamodb.Attribute(
                        name='timestamp',
                        type=dynamodb.AttributeType.NUMBER
                    )
                )
            ]
        )

        # Add TTL attribute
        cfn_table = self.table.node.default_child
        cfn_table.time_to_live_specification = dynamodb.CfnGlobalTable.TimeToLiveSpecificationProperty(
            enabled=True,
            attribute_name='ttl'
        )

        # Tags
        cdk.Tags.of(self.table).add('DR-Role', 'Global-Table')

        # Outputs
        cdk.CfnOutput(
            self,
            'TableName',
            value=self.table.table_name,
            description='DynamoDB table name'
        )
        cdk.CfnOutput(
            self,
            'TableArn',
            value=self.table.table_arn,
            description='DynamoDB table ARN'
        )